contracts/dao/neutron-flashloans/src/contract.rs

```diff
@@ -0,0 +1,403 @@
+    if ACTIVE_LOAN.load(deps.storage)?.is_some() {
+        return Err(FlashloanAlreadyActive {});
+    }
+
+    // Check that the amount does not have duplicate or zero coins
+    validate_amount(amount.clone())?;
+
+    // Getting the current balances of the source is necessary to make sure that the loan
+    // was returned, and the fee was paid.
+    let pre_loan_balances = get_pre_loan_balances(&deps, config.source.clone(), amount.clone())?;
+
+    // Calculate the fee to be paid by the borrower for receiving the loan
+    let fee = calculate_fee(amount.clone(), config.fee_rate)?;
+
+    // Calculate the expected balance of the source after the loan has been returned
+    // and the fee has been paid as (requested amount + fee).
+    let expected_balances = calculate_expected_balances(pre_loan_balances, fee.clone())?;
+
+    // Save all the information necessary to continue processing the loan request in the reply()
+    // handler.
+    ACTIVE_LOAN.save(
+        deps.storage,
+        &Some(ActiveLoan {
+            borrower: info.sender.clone(),
+            amount: amount.clone(),
+            fee,
+            expected_balances,
+        }),
+    )?;
+
+    // Send a (stargate -> authz -> bank) /cosmos.bank.v1beta1.MsgSend submessage with
+    // reply_on_success strategy (we want the transaction to be simply reverted in case of an
+    // error).
+    let msg_send = get_stargate_authz_bank_send_msg(env, config, info.sender, amount);
+    Ok(Response::new()
+        .add_submessage(SubMsg::reply_on_success(msg_send, AUTHZ_BANK_SEND_REPLY_ID))
+        .add_attribute("action", "execute_get_loan"))
+}
+
+// Check that the amount does not have duplicate or zero coins
+fn validate_amount(amount: Vec<Coin>) -> Result<(), ContractError> {
+    let mut denoms: HashSet<String> = HashSet::new();
+    for coin in amount {
+        if coin.amount.eq(&Uint128::zero()) {
+            return Err(ContractError::ZeroRequested { denom: coin.denom });
+        }
+
+        if denoms.contains(&coin.denom) {
+            return Err(ContractError::DuplicateDenoms {});
+        }
+
+        denoms.insert(coin.denom);
+    }
+
+    Ok(())
+}
+
+/// Returns the list of current balances on the contract's account for the coins that were
+/// requested by the borrower. If any actual coin balance is lower than the requested amount,
+/// returns an error.
+fn get_pre_loan_balances(
+    deps: &DepsMut,
+    source: Addr,
+    requested_amount: Vec<Coin>,
+) -> Result<Vec<Coin>, ContractError> {
+    // Prepare the query
+    let all_balances_query = BankQuery::AllBalances {
+        address: source.to_string(),
+    };
+    // Get the response (all balances)
+    let all_balances_response: AllBalanceResponse =
+        deps.querier.query(&all_balances_query.into())?;
+
+    // Filter all balances leaving only the balances of the requested coins
+    let mut pre_loan_balances: Vec<Coin> = vec![];
+    for requested_coin in requested_amount {
```

Comment:
Or i can get a gas error here.

TLDR infinite loops are not good

Reply:
I see your point, spamming might be an issue; will fix.